go-runtime/schema/extract_test.go

```diff
@@ -0,0 +1,324 @@
+package schema
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/TBD54566975/ftl/backend/schema"
+	"github.com/TBD54566975/ftl/internal/slices"
+	"github.com/alecthomas/assert/v2"
+)
+
+var symbols = []string{
+	"any", "bool", "[]byte", "float64", "int", "string", "time.Time", "ftl.Unit",
+}
+
+// generate all possible permutations for slices, maps and ftl.Option[...]
+func generatePermutations(symbols []string) []string {
+	permutations := append([]string{}, symbols...)
+	for _, symbol := range symbols {
+		permutations = append(permutations, "[]"+symbol)
+		permutations = append(permutations, "ftl.Option["+symbol+"]")
+		for _, nestedSymbol := range symbols {
+			permutations = append(permutations, fmt.Sprintf("map[%s]%s", symbol, nestedSymbol))
+		}
+	}
+	return permutations
+}
+
+func generateSymbolTypeStrings(symbols []string) map[string]string {
+	symbolTypeStringMap := make(map[string]string)
+	for _, symbol := range symbols {
+		symbolTypeStringMap[symbol] = getSchemaType(symbol).String()
+	}
+	return symbolTypeStringMap
+}
+
+func getSchemaType(symbol string) schema.Type {
+	switch symbol {
+	case "any":
+		return &schema.Any{}
+	case "bool":
+		return &schema.Bool{}
+	case "[]byte":
+		return &schema.Bytes{}
+	case "float64":
+		return &schema.Float{}
+	case "int":
+		return &schema.Int{}
+	case "string":
+		return &schema.String{}
+	case "time.Time":
+		return &schema.Time{}
+	case "ftl.Unit":
+		return &schema.Unit{}
+	default:
+		if strings.HasPrefix(symbol, "[]") {
+			return &schema.Array{Element: getSchemaType(symbol[2:])}
+		}
+		if strings.HasPrefix(symbol, "map[") {
+			idx := strings.Index(symbol, "]")
+			if idx == 5 {
+				// handle map[[]...]...
+				idx = strings.Index(symbol[6:], "]") + 6
+			}
+			key := getSchemaType(symbol[4:idx])
+			value := getSchemaType(symbol[idx+1:])
+			return &schema.Map{Key: key, Value: value}
+		}
+		if strings.HasPrefix(symbol, "ftl.Option[") {
+			return &schema.Optional{Type: getSchemaType(symbol[11 : len(symbol)-1])}
+		}
+		panic(fmt.Sprintf("unexpected symbol: %s", symbol))
+	}
+}
+
+func generateSourceCode(symbol string) string {
+	var codeBuilder strings.Builder
+
+	codeBuilder.WriteString("package test\n\n")
+	codeBuilder.WriteString(`import (` + "\n")
+	codeBuilder.WriteString(`	"context"` + "\n")
```

Comment:
Would it be better/easier to use raw string literals here instead of this `codeBuilder`? It seems a bit hard to write and update like this

Reply:
true! looks better